batch API: add suspended job

adtac opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**: See KEP: https://github.com/kubernetes/enhancements/pull/2234

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/2232

**Special notes for your reviewer**:

1. e2e tests require the alpha feature gate, so they won't run in presubmit until https://github.com/kubernetes/test-infra/pull/20833 is merged
2. integration tests coming in a follow-up PR

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Jobs API has a new .spec.suspend field that can be used to suspend and resume Jobs. This is an alpha field which is only honored by servers with the `SuspendJob` feature gate enabled.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/2234
```

cc @ahg-g @alculquicondor @soltysh

adtac commented:
/sig apps

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,9 +567,39 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+		} else if *job.Spec.Suspend {
+			// Job can be in the suspended state only if it is NOT completed.
+			exists := false
+			for _, c := range job.Status.Conditions {
+				if c.Type == batch.JobSuspended {
+					exists = true
+					break
+				}
+			}
```
lets have this in a `getSuspendedCondition` function

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,9 +567,39 @@ func (jm *Controller) syncJob(key string) (bool, error) {
 			now := metav1.Now()
 			job.Status.CompletionTime = &now
 			jm.recorder.Event(&job, v1.EventTypeNormal, "Completed", "Job completed")
+		} else if *job.Spec.Suspend {
```
we need to update the existing condition if there is one by flipping the status and updating the timestamps. I added the following to the KEP after discussing with reviewers:

"Note that when the suspend flag in the Job spec is flipped from true to false, the Job controller simply updates the existing suspend JobCondition status to false; it does not remove the condition or add a new one."

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,9 +567,39 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+				job.Status.Conditions = append(job.Status.Conditions, newCondition(batch.JobSuspended, "", ""))
+				jobConditionsChanged = true
+			}
+		} else {
```
we are in a a big nested if/else statement, if we don't want to refactor it in this PR, I suggest adding a comment beside each else. (e.g., in this case: `// Job not suspended`)

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,9 +567,39 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+				jobConditionsChanged = true
+			}
+		} else {
+			result := make([]batch.JobCondition, 0)
```
I am guessing this logic will change now that we expect a single instance of the condition. If we have a `getSuspendedCondition` function, you can reuse it here

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -727,6 +748,26 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
+	} else if active > parallelism {
+		diff := active - parallelism
+		errCh = make(chan error, diff)
+		jm.expectations.ExpectDeletions(jobKey, int(diff))
```
it seems you can set this inside the `deletePods` func?

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -727,6 +748,26 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
+
+	} else if active > parallelism {
+		diff := active - parallelism
+		errCh = make(chan error, diff)
```
shouldn't this be initialized inside the `deletePods` function?

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -727,6 +748,26 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
+		klog.V(4).Infof("job %q is suspended, deleting all %d active pods", jobKey, active)
+		active = 0
+		deletePods(numPodsToDelete)
+
```
remove empty line

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -695,20 +726,10 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
-		sort.Sort(controller.ActivePods(activePods))
-
-		active -= diff
+	deletePods := func(numPodsToDelete int32) {
```
I have refactored this into its own function :)

But I can resolve the conflicts later.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -917,15 +919,61 @@ func TestAddPodOrphan(t *testing.T) {
 	}
 }
 
+func TestSuspendJob(t *testing.T) {
```
Can we also test a Job that is created in suspended status?

adtac replied:
Done.

alculquicondor commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -93,6 +97,10 @@ func (jobStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object
 		newJob.Spec.TTLSecondsAfterFinished = nil
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && oldJob.Spec.Suspend == nil {
+		newJob.Spec.Suspend = nil
+	}
```
The feature is for the ability to suspend.

So, if the oldJob is suspended and the feature gets disabled, I think we should allow to resume

ahg-g replied:
right, but allowing it to resume means we shouldn't check the feature gate in the controller, which doesn't sound like a good idea either.

adtac replied:
if a job is suspended and the feature gate is disabled, when kube-controller-manager restarts, the job will automatically resume again

adtac replied:
So there are two cases when the feature gate is disabled:

1. resuming: this happens automatically as per my previous comment
2. suspending: I don't see why anyone would update from nil -> non-nil when the feature gate is disabled. That is basically no-op.

adtac replied:
Switched to bool instead of boolptr

alculquicondor commented on pkg/registry/batch/job/strategy_test.go:
```diff
@@ -88,8 +94,9 @@ func testJobStrategy(t *testing.T) {
 		},
 		Spec: batch.JobSpec{
 			Selector:                validSelector,
+			Suspend:                 newBool(true),
```
we should get rid of these methods and use the `pointer` package. But we can do that as followup for the entire file

adtac replied:
Changed it everywhere

alculquicondor commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -60,6 +60,16 @@ type JobList struct {
+	// This is an alpha field and requires enabling the SuspendJob feature gate.
+	// Defaults to false.
+	// +optional
+	Suspend *bool `json:"suspend,omitempty" protobuf:"varint,10,opt,name=suspend"`
```
Question: Why do we need a pointer instead of a plain boolean?

The answer might be: because we want webhooks to be able to differentiate whether a user set a preference. In that case, we shouldn't be filling up a default in `defaults.go`, but the strategy. Otherwise webhooks get the value already filled up.

But realistically, I think webhooks will only want to enforce suspend=true, regardless of user preference, so I don't see the need for a pointer.

ahg-g replied:
don't we also need this to drop the field when the feature is disabled?

alculquicondor replied:
then this again is the same question: what does it mean to disable the feature? https://github.com/kubernetes/kubernetes/pull/98727#discussion_r570467511

ahg-g replied:
it doesn't only mean disallowing the proposed behavior, but also disable the logic completely so that operators can shut it down if there is a bug or unexpected behavior in the initial implementation.

alculquicondor replied:
but in that case, the suspended=false doesn't have any alterations

adtac replied:
Resolved, see the other thread.

ahg-g commented on pkg/apis/batch/v1/defaults.go:
```diff
@@ -26,6 +26,10 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
 }
 
 func SetDefaults_Job(obj *batchv1.Job) {
+	if obj.Spec.Suspend == nil {
```
yes, but since the field is mutable, this case might be different depending on what we decide here: https://github.com/kubernetes/kubernetes/pull/98727#discussion_r570467511

lavalamp replied:
I think the feature flag should prevent new uses of the feature, but not clobber existing uses of the feature. I would not expect it to be checked in the default setter, since the default should be the same either way.

Feature flags should add safety, not put the system in a less tested 3rd state when turned off.

ahg-g replied:
ok, then I guess we need to do something different then in the update strategy. Currently the PR sets the field back to nil on the following condition ``` !utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && oldJob.Spec.Suspend == nil```, otherwise it allows the update. Since the field is defaulted here, all updates will be allowed because `oldJob.Spec.Suspend` will never be nil.

The other option is to not use a reference in the external type like `paused` in deployment and discussed in https://github.com/kubernetes/kubernetes/pull/98727/files#r571054554

adtac replied:
When the feature gate is on, we allow everything. When the feature gate is off, we have only a few combinations:

|old|new|what to do|
|-|-|-|
|nil|nil|we don't care, allow|
|nil|true/false|disallow ("prevent new uses of the feature")|
|true/false|nil|disallow ("not clobber existing uses of the feature" and "not put the system in a less tested 3rd state when turned off")|
|true/false|opposite|disallow|

So if the oldJob is nil, force newJob to stay nil. If the oldJob is non-nil, force newJob to be equal to oldJob. SG?

Updated strategy.go

alculquicondor replied:
Note that defaulting also happens when apiserver loads an object from storage. That means that old.Spec.Suspended is never nil (only false)

For that reason, I still suggest that we just get rid of the pointer and use a plain boolean.

When the feature gate is off, would the controller do anything different? Should we just control in the apiserver by disallowing transitions?

alculquicondor replied:
Again, the point of the feature gate is to restore the state of the cluster to a known working state, where .spec.suspended is always false. In that sense, I don't see a reason why we should prevent the true->false transition. We just need to prevent false->true.

adtac replied:
my only final issue with a non-pointer type is precedence. CronJob's suspend field is a boolptr too and has the same semantics from a API standpoint. it too is defaulted to false when loading from storage and the controller never sees a nil value.

@lavalamp for the final say? should booleans that default to false be bool pointers or plain bools? the controller using the field will never see a nil value thanks to defaulting.

adtac reviewed (commented):
addressed all but two comments, I'll respond to those shortly

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -915,3 +897,24 @@ func filterPods(pods []*v1.Pod, phase v1.PodPhase) int {
+	for i := range list {
+		if list[i].Type == cType {
+			matches = list[i].Status == status
+			if !matches {
```
`if matches`?

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -915,3 +897,24 @@ func filterPods(pods []*v1.Pod, phase v1.PodPhase) int {
+			if !matches {
+				list[i].Status = status
+				list[i].LastTransitionTime = metav1.Now()
+				isUpdated = true
```
I would just return here.
Also, it looks like `isUpdated` will always be true.

alculquicondor commented on pkg/registry/batch/job/strategy_test.go:
```diff
@@ -31,12 +31,9 @@ import (
 	_ "k8s.io/kubernetes/pkg/apis/batch/install"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/features"
+	utilpointer "k8s.io/utils/pointer"
```
just `pointer` if there is no conflict.

ahg-g commented on pkg/apis/batch/types.go:
```diff
@@ -88,6 +88,16 @@ type JobTemplateSpec struct {
+	// This is an alpha field and requires enabling the SuspendJob feature gate.
+	// Defaults to false.
+	// +optional
+	Suspend *bool
```
deployment uses `Paused`, so lets use `Suspended` here.

ahg-g commented:
Adi, it will be helpful if you split the API and implementation into separate PRs to simplify the API review. @lavalamp 
will be looking at it for us.

lavalamp commented:
> Adi, it will be helpful if you split the API and implementation into separate PRs to simplify the API review. @lavalamp
will be looking at it for us.

Actually if you just make the API and implementation separate commits in this PR, that's probably good enough. Since we prefer to merge it together anyway.

adtac reviewed (commented):
Split the PR main commit into two more commits: one for API change and one for controller changes.

adtac commented:
/retest

adtac commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

alculquicondor commented on test/e2e/node/ttlafterfinished.go:
```diff
@@ -68,12 +68,13 @@ func testFinishedJob(f *framework.Framework) {
 	ttl := int32(10)
 
-	job := e2ejob.NewTestJob("randomlySucceedOrFail", "rand-non-local", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit)
+	job := e2ejob.NewTestJob("randomlySucceedOrFail", "rand-non-local", v1.RestartPolicyNever, isSuspend, parallelism, completions, nil, backoffLimit)
```
I don't think an ever growing `NewTestJob` function is ideal. Specially since most tests don't care about the suspend flag.
Instead, just override the extra value where necessary, like it's done for ttl down below

ahg-g commented:
/triage accepted
/priority important-soon

ahg-g commented:
@lavalamp friendly ping

ahg-g commented:
/milestone v1.21

ahg-g commented:
/label api-review

lavalamp commented on staging/src/k8s.io/api/testdata/HEAD/batch.v1.Job.yaml:
```diff
@@ -43,6 +43,7 @@ spec:
       - 3_bQw.-dG6c-.x
     matchLabels:
       hjT9s-j41-0-6p-JFHn7y-74.-0MUORQQ.N4: 3L.u
+  suspend: true
```
I guess I expected only this change and not the others in these golden data files?

adtac replied:
I didn't expect this to change either, but these testdata changes are automatically generated by `UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test -run=TestCompatibility`. FWIW Aldo's PR's codegen (https://github.com/kubernetes/kubernetes/pull/98812) also generated identical changes to these files.

lavalamp replied:
OK

lavalamp commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -126,12 +126,22 @@ type JobSpec struct {
-	// The latest available observations of an object's current state.
-	// When a job fails, one of the conditions will have type == "Failed".
+	// The latest available observations of an object's current state. Can be
+	// one of "Suspended", "Failed", or "Complete".
```
Don't use wording that implies other conditions are not possible.

adtac replied:
I assume you meant "Don't use wording that implies other conditions are ~~not~~ possible" because these are the only three possible values. Updated to use "will be" instead of "can be" :+1:

alculquicondor replied:
I don't think that's what he meant. The wording shouldn't imply that there won't be other conditions in the future.

lavalamp replied:
Right, what Aldo said. :)

lavalamp commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -126,12 +126,22 @@ type JobSpec struct {
+	// controller. If a Job is suspended after creation (i.e. the flag goes from
+	// false to true), the Job controller will delete all active Pods associated
+	// with this Job. Users must design their workload to gracefully handle this.
+	// This is an alpha field and requires enabling the SuspendJob feature gate.
```
```suggestion
	// This is an alpha field and requires the SuspendJob feature gate to be enabled, otherwise it may not be set to true.
```

adtac replied:
Done.

lavalamp reviewed (commented):
I only skimmed the controller (edit: meaning, I reviewed everything else). Just a few nits.

ahg-g commented:
/assign @soltysh 

for the job controller changes

ahg-g commented:
> I only skimmed the controller (edit: meaning, I reviewed everything else). Just a few nits.

Thanks Daniel!

adtac reviewed (commented):
Thanks for the review Daniel!

lavalamp commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -126,12 +126,22 @@ type JobSpec struct {
-	// The latest available observations of an object's current state.
-	// When a job fails, one of the conditions will have type == "Failed".
+	// The latest available observations of an object's current state. Will be
+	// one of following types: "Suspended", "Failed", or "Complete".
```
No! It's an open ended list. We don't know what conditions other actors might add, nor do we know what future job controllers might say. It's critical that clients not take a hard dependency on this list never growing.

adtac replied:
Done.

soltysh commented on pkg/apis/batch/types.go:
```diff
@@ -149,6 +149,16 @@ type JobSpec struct {
+	// This is an alpha field and requires enabling the SuspendJob feature gate.
+	// Defaults to false.
+	// +optional
+	Suspend bool
```
I think this belong in the `api: add suspend field to Job` commit :wink:

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -727,7 +741,22 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
 		wait.Wait()
+	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && job.Spec.Suspend {
```
A new test case in `TestControllerSyncJob` should uncover all the edge cases, at least I'd like to see:
1. active job with some pods in various states being suspended (w/o required pod change, for example parallelism == activePods)
2. active job with some pods in various states being suspended (with pod change required, parallelism != activePods)
3. new job being suspended
4. deleted job

adtac replied:
>b/c it relies that a job actually needs a change

That's a great point, thank you for catching that! Isn't this achieved by checking for suspension state change in addition to jobNeedsSync?

>There's a question what happens if start time isn't set (a new suspended job) should we set it or keep it nil until it actually starts?

We decided in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2232-suspend-jobs#design-details) that:

>If a Job is created with `suspend: true`, the `StartTime` field of the Job status is set only when it is resumed for the first time.

I agree with @ahg-g about ActiveDeadlineSeconds. The way I've updated the code, the timer will now only be set after StartTime is set, which will happen only after the job is unsuspended for the first time i.e. jobs created in the suspended state can stay that way for longer than ActiveDeadlineSeconds.

>Parallelism is a mutable field, and users can set this field back to 0; what I failed to understand when reading the code is how a job update that sets parallelism=0 result in a job sync (i.e., a call to manageJob) because I assume we can do the same for when suspend is changed, and so we keep the code pretty much as is.

@ahg-g Setting parallelism=0 will normally trigger the `active > parallelism` condition below, but we get a similar behaviour by checking for suspend before that condition. The suspend branch (this branch) does the same thing as the `active > parallelism` branch when `parallelism` is 0 through the `deletePods` function. Functionally they're similar, but they're simply executed in different branches so that the log message etc. could be different. Unless I misunderstood your question?

soltysh replied:
I had a chat with all the parties in the slack, the confusion was coming from my initial different take on how we should approach job suspension, which I thought will be like a hammer, but we eventually decided it will not, but it'll be treated equally with any other change to a job spec (like @ahg-g mentioned above parallelism, for example). 

So the only remaining thing is changes wrt ADS and unit tests.

adtac replied:
Done.

soltysh reviewed (changes requested):
The controller needs significant changes.

adtac commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -126,6 +126,16 @@ type JobSpec struct {
+	// This is an alpha field and requires enabling the SuspendJob feature gate.
+	// Defaults to false.
+	// +optional
+	Suspend bool `json:"suspend,omitempty" protobuf:"varint,10,opt,name=suspend"`
```
Moved it there :)

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -702,22 +729,26 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
-	}
-
-	if active < parallelism {
+	} else if active > parallelism {
```
in the new logic, we check for removals even if `active <= parallelism`. This is because there could be more than one pod with the same index, and we want to remove one of them.

I'll add a test for that.

alculquicondor replied:
Created #99789

alculquicondor replied:
I thought this wouldn't pass my test now

adtac replied:
Correct, fixed

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -910,3 +941,31 @@ func errorFromChannel(errCh <-chan error) error {
+		}
+	}
+	if !matches && status != v1.ConditionFalse {
+		return append(list, newCondition(cType, status, "", "")), true
```
shouldn't we have a reason and message?

adtac replied:
I can see how reason/message makes sense for job failures, so I couldn't think of anything useful for reason/message for a job getting suspended, but I've added something anyway, PTAL :)

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -910,3 +941,31 @@ func errorFromChannel(errCh <-chan error) error {
+				matches, isUpdated = true, true
+			}
+			// At this point, `matches` is always true. `isUpdated` may be false.
+			break
```
why not just return here?

soltysh replied:
Probably just less return points, I'm fine.

alculquicondor replied:
I think all the `matches`, `isUpdated` changes is rather harder to follow. But nit.

adtac replied:
I've updated the function and added a bunch of unit tests, lmk if this reads better

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -705,7 +867,7 @@ func TestSyncJobPastDeadline(t *testing.T) {
 
 			// job & pods setup
-			job := newJob(tc.parallelism, tc.completions, tc.backoffLimit, batch.NonIndexedCompletion)
+			job := newJob(false, tc.parallelism, tc.completions, tc.backoffLimit, batch.NonIndexedCompletion)
```
I said this early, and I don't know why it got lost.
Rather than adding more arguments, let's just modify the job after `newJob` for the tests that need it. Honestly, I should have done the same for completion mode.

soltysh replied:
yes :100:

adtac replied:
I think we discussed this in the e2e test instead last time haha, done.

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,10 +571,32 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+			if job.Spec.Suspend {
+				// Job can be in the suspended state only if it is NOT completed.
+				var isUpdated bool
+				job.Status.Conditions, isUpdated = ensureJobConditionStatus(job.Status.Conditions, batch.JobSuspended, v1.ConditionTrue)
```
```suggestion
				job.Status.Conditions, jobConditionsChanged = ensureJobConditionStatus(job.Status.Conditions, batch.JobSuspended, v1.ConditionTrue)
```

ahg-g replied:
and remove `isUpdated` and `jobConditionsChanged = true` below

alculquicondor replied:
what if `jobConditionsChanged` was true before calling `ensureJobConditionStatus`?

ahg-g replied:
is that possible?

alculquicondor replied:
you're right, it's not

ahg-g replied:
yeah, ideally we should be trying to set the condition in one place in a sync cycle.

adtac replied:
Consider a situation where a job failed in this syncJob cycle (jobConditionsChanged set to true in the Failed branch) but the job 
has never been suspended. Then ensureJobConditionStatus's second return val will be false because the job has never been suspended (this is normal), but this would overwrite the JobFailed condition changed status so the controller won't write the update.

Or do you mean changing only the suspend=true branch (which would be the worst of both worlds IMO)? I genuinely think explicitly receiving isUpdated and updating jobConditionsChanged separately is cleaner.

ahg-g replied:
> Or do you mean changing only the suspend=true branch (which would be the worst of both worlds IMO)? I genuinely think explicitly receiving isUpdated and updating jobConditionsChanged separately is cleaner.

I mean the exact change I am proposing, nothing more. It is a nit, so up to you, but I find it redundant the way it is done.

ahg-g replied:
> > Consider a situation where a job failed in this syncJob cycle (jobConditionsChanged set to true in the Failed branch) but the job has never been suspended. Then ensureJobConditionStatus's second return val will be false because the job has never been suspended (this is normal), but this would overwrite the JobFailed condition changed status so the controller won't write the update.

If you get into that branch, you will never reach here.

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,10 +571,32 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+			} else {
+				// Job not suspended.
+				var isUpdated bool
+				job.Status.Conditions, isUpdated = ensureJobConditionStatus(job.Status.Conditions, batch.JobSuspended, v1.ConditionFalse)
```
```suggestion
				job.Status.Conditions, jobConditionsChanged = ensureJobConditionStatus(job.Status.Conditions, batch.JobSuspended, v1.ConditionFalse)
```

adtac replied:
See above comment.

adtac reviewed (commented):
Thanks for the review, @soltysh @agh and @alculquicondor. I've updated everything except for the StartTime thing which I disagree with (see slack). Pushing in a sec after a rebase :)

adtac commented on pkg/controller/job/job_controller.go:
```diff
@@ -486,10 +486,11 @@ func (jm *Controller) syncJob(key string) (bool, error) {
-	if job.Status.StartTime == nil {
+	// Job first start. Set StartTime and start the ActiveDeadlineSeconds timer
+	// only if the job is not in the suspended state.
+	if job.Status.StartTime == nil && !suspended {
```
discussed this on slack, we've agreed to reset StartTime every time a Job is resumed because: (1) it keeps the code simpler, (2) resuming a job created in the suspended state sets StartTime only when it's resumed, so this would offer a similar behaviour

ahg-g commented:
@lavalamp the API changes have been addressed, PTAL; we are currently just tuning some details related to the controller implementation.

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -660,9 +695,12 @@ func pastBackoffLimitOnFailure(job *batch.Job, pods []*v1.Pod) bool {
+// return false.
 func pastActiveDeadline(job *batch.Job) bool {
-	if job.Spec.ActiveDeadlineSeconds == nil || job.Status.StartTime == nil {
+	suspended := utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && job.Spec.Suspend != nil && *job.Spec.Suspend
```
create a `jobSuspended` function and use it directly in the if statement below, ditto for other places.

adtac replied:
Thanks, done.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -702,22 +740,27 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
-	}
-	podsToDelete := activePodsForRemoval(job, activePods, int(rmAtLeast))
-	if len(podsToDelete) > 0 {
+	suspended := utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && job.Spec.Suspend != nil && *job.Spec.Suspend
```
have a function for this, as it's used more than once.

lavalamp commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -166,21 +169,38 @@ type JobSpec struct {
+	// a Job is suspended, one of the conditions will have type "Suspended" and
+	// status true; when the Job is resumed, the status of this condition will
+	// become false. When a Job is completed, one of the conditions will have
+	// type "Complete" and status true.
```
Looks good now.

I'm not looking at the implementation here, but please note this line from our [condition guidance](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties):

> Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown. This allows other components in the system to know that the condition exists ...

alculquicondor replied:
That hasn't been respected so far :(

Should we fix that?
I feel like we can potentially break people that might just be relying on the existence of the condition without checking the status.

lavalamp replied:
Yes it should probably be fixed, perhaps not in this PR.

Checking for the condition but not its value doesn't seem like a super plausible mistake to me--hopefully I'm not wrong about that haha

lavalamp commented:
/approve

for the api.

ahg-g commented:
/hold

to prevent accidental merge since this is now approved.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -185,8 +201,17 @@ func TestControllerSyncJob(t *testing.T) {
 		expectedConditionReason string
 		expectedCreatedIndexes  sets.Int
 
+		// expectations after update
```
Uhm... I'm not liking how we are complicating the implementation of this test.

Before this change, it is basically testing one `syncJob` call. What if you instead have it as: `wasSuspended` and `suspend`. Then you prep the job adding the corresponding condition before calling sync.

And then you still have `TestsSuspendJob` below to test more specific things that don't apply to this testtable.

adtac replied:
>What if you instead have it as: wasSuspended and suspend.

This way, we can't test the effects of suspend at Job creation *and* update, which is required for some of the tests (e.g. create a Job suspended, make sure it doesn't create pods, update jobs, make sure pods are created now). I don't see any way to do this without a update to the existing Job.

And I considered re-using TestSuspendJob but there would be so much repeated code in the stage before the update that I think it's not worth it. I agree that testing updates in this function makes it more complicated, but so is the job controller now :)

I'm open to suggestions to refactor the tests in a separate PR if it's possible to do it without too much code repetition though (after the code freeze but before the tests freeze if that's okay).

alculquicondor replied:
> This way, we can't test the effects of suspend at Job creation and update, which is required for some of the tests (e.g. create a Job suspended, make sure it doesn't create pods, update jobs, make sure pods are created now)

That can be seen as 2 separate tests.

> I don't see any way to do this without a update to the existing Job.

In general, a sync doesn't know what the previous sync did, so they are independent. Only thing it can see is what's running and what is the past status. So you just have to modify those.

soltysh commented on pkg/apis/batch/types.go:
```diff
@@ -187,19 +190,36 @@ type JobSpec struct {
+	// requires the SuspendJob feature gate to be enabled; otherwise this field
+	// may not be set to true. Defaults to false.
+	// +optional
+	Suspend *bool
```
:+1:

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,10 +571,40 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+			if job.Spec.Suspend != nil && *job.Spec.Suspend {
+				// Job can be in the suspended state only if it is NOT completed.
+				var isUpdated bool
+				job.Status.Conditions, isUpdated = ensureJobConditionStatus(job.Status.Conditions, batch.JobSuspended, v1.ConditionTrue, ".spec.suspend is true", "Job suspended")
```
Correct, we already do that in this controller, see https://github.com/kubernetes/kubernetes/blob/269d62d895c29743931bfaaec6e8d37ced43c35f/pkg/controller/job/job_controller.go#L519

adtac replied:
Thanks, done.

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -566,10 +571,40 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+					// (ActiveDeadlineSeconds is interpreted as the number of seconds a
+					// Job is continuously active.)
+					now := metav1.Now()
+					job.Status.StartTime = &now
```
:+1:

soltysh reviewed (approved):
/approve 
controller bits

Commit:
batch: add suspended job

Signed-off-by: Adhityaa Chandrasekar <adtac@google.com>

adtac commented:
@soltysh @lavalamp @alculquicondor @ahg-g Thanks for the review once again. I believe I've addressed all review comments and tests should pass. Please take one last look :)

adtac commented:
@lavalamp the TestRoundTripTypes unit test failure seems to be due to defaulting suspend to &false (see below). Do you recommend not defaulting since the job controller already checks for nil values with something like `suspend != nil && *suspend`?

```
        -                               Suspend:                 nil,
        +                               Suspend:                 &false,
```

desponda commented:
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day, tomorrow the 9th! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

alculquicondor commented:
>  Do you recommend not defaulting since the job controller already checks for nil values with something like suspend != nil && *suspend?

It's better to be explicit in the API.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1162,6 +1359,98 @@ func TestAddPodOrphan(t *testing.T) {
 	}
 }
 
+func TestSuspendJob(t *testing.T) {
```
I'm not following. What is this testing? Please add a comment.

adtac replied:
It tests that updating suspend from true to false (or false to true) requeues the job for an update so that the controller eventually observes the change, but this is now being tested in the syncJob test as well, so this test is now redundant. Thanks for catching that, removed.

lavalamp commented:
I agree that suspend should be defaulted false.

adtac commented:
I've solved the unit test issue. The problem is that we're fuzzing the internal JobSpec type, not the v1 type, so fuzzing the nil value doesn't make sense because the internal type is always after defaults are applied, so it will never be nil. The roundtrip test was failing because of this. cc @lavalamp 

Thanks to @alculquicondor for helping with the debug.

ahg-g commented:
> @soltysh @lavalamp @alculquicondor @ahg-g Thanks for the review once again. I believe I've addressed all review comments and tests should pass. Please take one last look :)

Thanks Adi, looks good to me.

ahg-g commented:
@adtac I think we are good to squash

adtac commented:
ack, squashed

alculquicondor reviewed (commented):
/lgtm

adtac commented:
/hold cancel
API and sig-apps approved, please feel free to hold again

ahg-g commented:
/lgtm

alculquicondor commented:
/lgtm

alculquicondor commented:
for future reference: keeping this unsquashed was fine

alculquicondor commented:
/retest

alculquicondor commented:
/retest
